Replace Google Charts with Chart.js

fanthos opened the pull request:
I am opening this PR for discuss some detailed information about how to implement the charts with working code.
Currently, I am using Billboard.js and D3, they costs about 380kb when build as Polymer fragments and 120kb gzipped. I am not sure if there is other way to reduce the size(compared with 300kb+ gzipped Google Charts).
The timeline chart is modified from https://github.com/jjagielka/d3-gtimeline .

fanthos commented:
Related issue: #349

andrey-git commented:
Is all of this JS HA-specific or is it a fork of another code?

fanthos commented:
https://github.com/fanthos/home-assistant-polymer/blob/baa6b79ccaf3a5a4bcffc623f2dc1b54626a415f/src/components/entity/ha-chart-base.html#L48-L437
These code forked from [d3-gtimeline](https://github.com/jjagielka/d3-gtimeline) under BSD and modified for HA. Other code are HA specific.
Bower dependency Billboardjs licensed under MIT.

balloob commented:
If you fork a library, please do not include it in our codebase but instead publish it to NPM instead.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,514 @@
+    </style>
+    <div id="chartTarget"></div>
+  </template>
+  <script src="../../../bower_components/d3/d3.min.js"></script>
```
These should not be inside the dom-module.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,514 @@
+/* global d3 bb */
+  /**
+ * Original: d3-gtimeline at Github
+ * Modified and rewrited for Home Assistant
```
Why was it modified and rewritten? Would it be possible to contribute back to the original library the functionality that we need?

fanthos replied:
Some feature can push back to the original one, but the data structure is changed. I moved the chart code to standalone Github repo.

balloob commented on src/components/entity/ha-chart-line.html:
```diff
@@ -0,0 +1,142 @@
+</dom-module>
+<script>
+
+Polymer({
```
Please use new Polymer 2 class syntax

fanthos replied:
Will merge with history-chart-line later.

balloob commented on src/components/entity/ha-chart-line.html:
```diff
@@ -0,0 +1,142 @@
+Polymer({
+  is: 'ha-chart-line',
+
+  useNativeCSSProperties: true,
```
I thought that native was being used on supported platforms. Why toggle this? We shouldn't do this.

balloob commented on src/components/entity/ha-chart-line.html:
```diff
@@ -0,0 +1,142 @@
+
+      stateInfo.forEach(function (state) {
+        var timeStamp = new Date(state.last_changed);
+        // if (timeStamp > endTime) {
```
Stale comment

balloob commented on src/components/entity/ha-chart-timeline.html:
```diff
@@ -0,0 +1,172 @@
+</dom-module>
+<script>
+
+Polymer({
```
Please use Polymer 2 class syntax.

fanthos commented:
Right now bower_components/billboard.js wouldn't build with script/build_frontend because babel is trying to compile billboard.js. Currently I am using prebuilt version located in `billboard.js/dist`, so I think that I don't have to rebuild the dependency with babel. Not sure what I should do with it.
```
[17:17:42] Using gulpfile /mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/gulpfile.js
[17:17:42] Starting 'default'...
[17:17:42] Starting 'clean'...
[17:17:42] Finished 'default' after 4.9 ms
[17:17:42] Finished 'clean' after 24 ms
[17:17:42] Starting 'run_rollup'...
[17:17:46] Finished 'run_rollup' after 4.12 s
[17:17:46] Starting 'ru_all'...
[17:17:46] Finished 'ru_all' after 1.13 ms
[17:17:46] Starting 'build'...

events.js:182
      throw er; // Unhandled 'error' event
      ^
ReferenceError: Unknown plugin "add-module-exports" specified in "/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/bower_components/billboard.js/.babelrc" at 0, attempted to resolve relative to "/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/bower_components/billboard.js"
    at /mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:180:17
    at Array.map (native)
    at Function.normalisePlugins (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
    at OptionManager.mergeOptions (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
    at OptionManager.init (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
    at File.initOptions (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/index.js:212:65)
    at new File (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/file/index.js:135:24)
    at Pipeline.transform (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
    at DestroyableTransform._transform (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/gulp-babel/index.js:36:22)
    at DestroyableTransform.Transform._read (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_transform.js:182:10)
    at DestroyableTransform.Transform._write (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_transform.js:170:83)
    at doWrite (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_writable.js:406:64)
    at writeOrBuffer (/mnt/c/Workspaces/repos/github/home-assistant/homeassistant/components/frontend/www_static/home-assistant-polymer/node_modules/readable-stream/lib/_stream_writable.js:395:5)
```

balloob commented:
Can be fixed by disabling the use of `babelrc` in the build script. On this line: https://github.com/home-assistant/home-assistant-polymer/blob/master/gulp/common/transform.js#L14 add `babelrc: false,`

fanthos commented:
Not sure if this change will affect other denpendency, will check tomorrow.

fanthos commented:
Seems works as excepted

fanthos commented:
I think the code can consider finished.
The houndci keep failing but I don't know the reason.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+<link rel='import' href='../../../bower_components/polymer/polymer.html'>
```
this needs to import polymer-element.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+    <div id="chartTarget"></div>
+  </template>
+</dom-module>
+<script src="../../../bower_components/d3/d3.js"></script>
```
Note to self to not merge this until I have reworked this to be loaded on demand.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+   * @customElement
+   * @polymer
+   */
+  class ExtenedChart extends Polymer.Element {
```
we should name classes after the custom element name, so in this case HaChartBase

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+/* global d3 bb */
+
+(function () {
+  if (customElements.get('ha-chart-base')) return;
```
This makes no sense, if it should be only included once it should be only loaded once.c

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+      this.connected = true;
+    }
+    onPropsChange() {
+      // if (!this._cssLoaded)
```
stale comment

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+      // if (!this._cssLoaded)
+      //     return;
+      if (!this.connected) return;
+      if (!this.publish || !this.data || !this.data.data) return;
```
Please merge with previous statement.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+      };
+      chartData.bindto = this.$.chartTarget;
+      if (!chartData.data.onclick) {
+        // debugger;
```
stale comment

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+        }
+      };
+      chartData.bindto = this.$.chartTarget;
+      if (!chartData.data.onclick) {
```
This will never be set because you initialize `chartData.data` to not have this attribute.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+      chartData.bindto = this.$.chartTarget;
+      if (!chartData.data.onclick) {
+        // debugger;
+        chartData.data.onclick = (dataPoint) => {
```
Any reason this cannot be included in the object definition above?

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,182 @@
+          type: Object,
+          observer: 'onPropsChange'
+        },
+        newData: {
```
This makes no sense, why not use `data` ? 

Also, we should never call anything new, it is bound to be not the new thing in the future.

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -1,20 +1,15 @@
 
+<dom-module id='state-history-chart-line'>
+  <template>
+      <ha-chart-base publish data="[[chartData]]" selected-element ="{{selectedDataPoint}}"></ha-chart-base>
```
selected-element is a read-only property. This should not work ?

balloob replied:
selectedDataPoint is not being used nor defined

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -138,21 +129,24 @@
-        var data = [];
-        var dataTable = new window.google.visualization.DataTable();
-        // array containing [time, value1, value2, etc]
+        // array containing [value1, value2, etc]
```
stale comment.

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -179,68 +181,60 @@
 
-        dataTable.addRows(data);
-        return dataTable;
+        Array.prototype.push.apply(dataTables, data);
```
`dataTables` is an array, no reason to use `Array.prototype`

balloob commented on src/components/state-history-chart-timeline.html:
```diff
@@ -1,11 +1,11 @@
+
+<dom-module id='state-history-chart-timeline'>
+  <template>
+      <ha-chart-base publish data="[[chartData]]" selected-element ="{{selectedDataPoint}}"></ha-chart-base>
```
selectedDataPoint not being used/defined.

balloob commented on src/components/state-history-chart-timeline.html:
```diff
@@ -41,39 +41,22 @@
-    function addRow(entityDisplay, stateStr, start, end) {
-      var stateDisplay = stateStr.replace(/_/g, ' ');
-      dataTable.addRow([entityDisplay, stateDisplay, start, end]);
+    if (!stateHistory || stateHistory.length === 0) {
```
Can merge this if with the one above.

balloob replied:
Actually, this check is now wrong because you removed the code that made the div all empty. Now if the new data is empty, it will not remove the old one and thus show stale data

balloob commented:
This PR has conflicts now 😞

fanthos commented:
I will look into this later

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.

fanthos commented:
I had more work on this PR, it seems work now.
Should I re-open the PR or open a new PR?

fanthos commented:
@balloob 
I am not able to reopen this PR.
Should I open a new PR or someone can reopen this one?

homeassistant commented:
Hello @fanthos,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ee2b9c5122efbf3278a4cad5f18cbe9b37766fe0`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;ee2b9c5122efbf3278a4cad5f18cbe9b37766fe0) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`017815141150f5749a0b6faa6cf99c0fc08bcecf`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;017815141150f5749a0b6faa6cf99c0fc08bcecf) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`7b61826134edb0e1a0e93d0346009478eb7f6ba5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;7b61826134edb0e1a0e93d0346009478eb7f6ba5) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.
- [`7d86f30a6b30eb2174a56d03cd5ea4bf500ab03d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant-polymer&#x2F;pull&#x2F;429&#x2F;commits&#x2F;7d86f30a6b30eb2174a56d03cd5ea4bf500ab03d) - This commit has something that looks like an email address (dllcadugmiet@web.de). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

fanthos committed:
chartjs test

fanthos committed:
[WIP] Modified for Chart.js

fanthos committed:
Tweaking styles ( tooltips and lines )

fanthos committed:
Almost done
TODO:
Change tooltips to HTML tag
Improve color function

fanthos committed:
More work on Tooltips

fanthos committed:
Merge branch 'master' into dev

fanthos committed:
Improve update logic
Fix linting

fanthos committed:
resolve conflict

andrey-git commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,228 @@
+      },
+      connected: {
+        type: Object,
+        readOnly: false,
```
I don't think `readOnly: true` has any effect.

andrey-git replied:
Also, is this supposed to be a boolean?

andrey-git commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,228 @@
+      connected: {
+        type: Object,
+        readOnly: false,
+        observer: 'onPropsChange'
```
Make a single observer that will observe all the variables instead. Otherwise if both `data` and `connected` change together the observer would be called twice.

andrey-git commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,228 @@
+      },
+    };
+  }
+  connectedCallback() {
```
call `super.connectedCallback()`

andrey-git commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,228 @@
+  connectedCallback() {
+    this.connected = true;
+    this.addEventListener('iron-resize', () => {
+      this.async(this.resizeChart, 10);
```
I don't think Polymer2 has `this.async` ...

andrey-git reviewed (commented):
Is this ready for review?

fanthos committed:
[WIP]
Create new tooltip mode hack.
Add axis padding to top and botton to prevent axis cutoff

fanthos committed:
Merge branch 'dev' of https://github.com/fanthos/home-assistant-polymer into chartjs

fanthos committed:
TODO: cleanup

fanthos commented:
No idea with how to fix this issue:
![image](https://user-images.githubusercontent.com/1490719/34245075-55dbd048-e663-11e7-8b04-a751c070d87a.png)

fanthos committed:
FIXME: tooltip in history graph not working correctly
reorganize some code

fanthos committed:
Merge branch 'master' of https://github.com/home-assistant/home-assistant-polymer into chartjs

fanthos committed:
Merge remote-tracking branch 'remotes/upstream/master' into chartjs

fanthos committed:
fix build problem

fanthos committed:
Merge branch 'master' of https://github.com/home-assistant/home-assistant-polymer into chartjs

fanthos committed:
Fix color and tooltip issue
Fix label max width for timeline chart

fanthos committed:
Merge branch 'dev' of https://github.com/fanthos/home-assistant-polymer into chartjs

# Conflicts:
#	src/components/entity/ha-chart-base.html

fanthos committed:
update dep

fanthos commented:
I think this PR is ready for review.

michaelarnauts reviewed (commented):
There is still some commented out code that should be cleaned up. Thanks for the effort!

fanthos committed:
Fix strange color after build due to `uglify` bug with reference the minified version.
Make line chart behavior more similar to Google Charts.
Make the chart honor to `unknown` and other state by manually calculate point value.

andrey-git reviewed (commented):
![chartsjs](https://user-images.githubusercontent.com/5478779/34455270-24399db8-ed84-11e7-9171-bea9081b36d6.png)

In timeline mode the name is cut.
Is it possible to put the state (For example "on") in the timeline chart like in google charts?

andrey-git commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,415 @@
+    </div>
+  </template>
+</dom-module>
+<script src="../../../bower_components/moment/moment.js"></script>
```
edit: Actually just add it to bower.json, as timeline doesn't work with minimized momentsjs

andrey-git replied:
Actually those should be loaded on demand. Putting them in script.src will pull them into frontend.html

fanthos replied:
I think the whole ha-chart-base should load on demand because currently only Chart.js depends on moment.js.

fanthos replied:
I didn't test timeline chart with the bundled version.

armills commented:
Also, how large is the included moment.js? I know we've tried to avoid it in the past due to the file size.

andrey-git commented:
Moment is 130k, 52k minimized.

In any case those should be loaded on demand and not bundled into frontend.html

andrey-git commented on src/components/state-history-chart-line.html:
```diff
@@ -1,265 +1,269 @@
-        options.chartArea.left = 40;
-        options.chartArea.height = '80%';
-        options.chartArea.top = 5;
+      function addColumn(nameY, type) {
```
I think there is no need to do string manipulations - just pass fill and step as params.

andrey-git commented on src/components/state-history-chart-line.html:
```diff
@@ -1,265 +1,269 @@
+        processState = function (state) {
+          if (!state.attributes) return;
+          var curTemp = safeParseFloat(state.attributes.current_temperature);
+          var heating = state.state === 'heat' ? curTemp : null;
```
I think we should add the fill for both 'heat
 and 'cool' as separate columns.

However maybe we should check if at least one datapoint exists - otherwise don't add the column.

andrey-git reviewed (commented):
Is it possible to show legend, like in google charts?
I think it is useful for charts with a few lines, like climate

fanthos commented on bower.json:
```diff
@@ -55,7 +55,9 @@
-    "webcomponentsjs": "^1.0.10"
+    "webcomponentsjs": "^1.0.10",
+    "chart.js": "~2.7.1",
+    "chartjs-chart-timeline": "fanthos/chartjs-chart-timeline#^0.1.1"
```
I think the advantage is easier to use it on other projects. The disadvantage is if there's some bug, people may need to fix it in two repo.

fanthos commented:
Line chart has legend support, so it can be enabled.

balloob commented:
Yeah we should make sure we load the chart JavaScript and moment.js on demand, they are big. I don't mind adding the code to load it on demand after this PR is merged.

andrey-git commented:
I think on-demand load is a must. If we merge this PR without on-demand load we will hurt load performance for everyone for sake of better graphs.

fanthos committed:
fix bugs

fanthos committed:
Merge branch 'dev' of https://github.com/fanthos/home-assistant-polymer into chartjs

# Conflicts:
#	src/components/state-history-chart-line.html

fanthos commented:
fixed

fanthos committed:
Remove label for only one data in timeline chart.
Fix bug for infinite loop in some cases

fanthos committed:
Add HTML legend to chart.

fanthos committed:
Fix isSingleDevice bug due to calculation.
Add isSingleDevice property support.

fanthos committed:
fix for lint

fanthos committed:
Replace innerHTML code with polymer node.

fanthos committed:
Replace tooltip with HTML code

fanthos committed:
fix tooltip style

fanthos committed:
move default tooltip mode to plugin

fanthos commented:
I think this PR can consider ready. Please let me know if I missed something.

fanthos committed:
Resize seems already supported by chartjs.

Merge branch 'master' of https://github.com/home-assistant/home-assistant-polymer into chartjs

# Conflicts:
#	src/components/state-history-chart-line.html
#	src/components/state-history-chart-timeline.html
#	src/components/state-history-charts.html

fanthos committed:
LINTING

fanthos committed:
fix
Move localize history data to Timeline Chart.
Fix timeline static color.
Rework on chart resize.

fanthos committed:
Merge remote-tracking branch 'remotes/upstream/master' into chartjs

# Conflicts:
#	src/components/state-history-charts.html

fanthos committed:
Bug fix:
Chart may disappear on some case.
Timeline chart calculation issue.
Change timeline chart hidden logic.

fanthos committed:
fix tooltip
rework for resize event

fanthos committed:
lint

balloob commented on src/components/state-history-chart-timeline.html:
```diff
@@ -1,85 +1,66 @@
-<link rel="import" href="../../bower_components/polymer/polymer-element.html">
+<link rel='import' href='../../bower_components/polymer/polymer.html'>
```
This should still be polymer-element (that's what the class extends from)

fanthos committed:
element

balloob commented on bower.json:
```diff
@@ -55,7 +55,9 @@
     "web-animations-js": "^2.2.5",
-    "webcomponentsjs": "^1.0.10"
+    "webcomponentsjs": "^1.0.10",
+    "chart.js": "~2.7.1",
```
You need to add moment too for Chart.js to work.

balloob commented:
![image](https://user-images.githubusercontent.com/1444314/35947544-342389e8-0c1d-11e8-8c71-c821047ef6c5.png)

Font inside the timeline is using a serif font. Can we change that to Roboto (or at least a sans-serif font)?

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,529 @@
+// Use minified(Chart.min.js) version to fix strange color after uglify
+// eslint-disable-next-line no-unused-vars
+/* global Chart moment Color */
+(function () {
```
Instead of executing it inside an IIFE, you can just wrap it inside a block:

```js
{
  Chart.Interaction.modes.neareach = …
}

```

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,529 @@
+/* global Chart moment Color */
+(function () {
+  Chart.Interaction.modes.neareach = function (chart, e, options) {
+    var getRange = {
```
We're using es2015 so we should use const and let.

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,529 @@
+        }
+      }
+    }
+    var ret = elements.filter(function (n) { return n !== undefined; });
```
Use a lambda: `return elements.filter(n => n !== undefined)`

balloob commented on src/components/entity/ha-chart-base.html:
```diff
@@ -0,0 +1,529 @@
+    var options = this.data;
+    var data = options.data;
+
+    if (this.isTimeline) {
```
Use a guard clause to keep the code readable

```js
if (!this.isTimeline) {
  this._chart.resize();
  return;
}

if (data.datasets.length === 0) return;

const axis = …

```

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -1,299 +1,302 @@
-    }
-
-    return result;
+'use strict';
```
Only works inside functions. You can remove it.

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -1,299 +1,302 @@
-          type: Object,
-        },
+  drawChart() {
+    var unit = this.unit;
```
`const`

balloob commented on src/components/state-history-chart-line.html:
```diff
@@ -1,299 +1,302 @@
-      this._isAttached = false;
-      window.removeEventListener('resize', this._resizeListener);
-    }
+    // startTime = new Date(Math.min.apply(null, deviceStates.map(function (states) {
```
stale comment?

balloob commented:
Alright, sorry a code review took so long. Looking at graphing code is not my favorite 😉 

Gist of my comments are:
 - Get rid of the usage of `var`, use `const` and `let` instead.
 - Use lambda functions (`=>`) as much as possible
 - Move all chart/moment imports and the `nearest` mode to a separate HTML file that is then imported. I will make sure to extract that HTML import and have it be loaded on demand.

balloob commented:
I'll make sure to code review it again as soon as possible once the fixes have been done.

I want to aim to get this in before release 0.64. That is going to be released in 2.5 weeks.

fanthos committed:
Replace `var` to `let`.
Move import and ChartJs injection code to `ha-chart-scripts.html`.

fanthos commented:
Thanks.
I moved the scripts to a separated file but don't know which folder should I put it.

fanthos committed:
lint: convert more let to const

balloob commented:
You can add it to `src/resources`.

balloob commented:
Note that the tests also have detected an error of an undefined property.

fanthos committed:
fix font
fix undef

fanthos committed:
update bower.json

fanthos committed:
move

balloob reviewed (approved):
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

balloob commented:
I will update it tomorrow to do the code splitting. Thanks for your patience and hard work!

andrey-git commented:
Was on-demand loading implemented?

balloob commented:
I will add that tomorrow

balloob committed:
Load chart code on demand

balloob commented:
Added 👍

balloob commented:
Just realized this PR has been 6 months in the making.

balloob commented:
Verified this PR works with both ES5 and latest in release mode. Niceeeee.

andrey-git commented on script/build_frontend:
```diff
@@ -27,6 +27,10 @@ cp build/panels/*.html $OUTPUT_DIR/panels
 mkdir $OUTPUT_DIR_ES5/panels
 cp build-es5/panels/*.html $OUTPUT_DIR_ES5/panels
 
+# Chart code
```
Don't we need to fingerprint it?

balloob replied:
I can't wait to Polymer 3 and native code splitting support via rollup/Webpack. No more hacky scripts on our part

andrey-git commented:
Why is the legend generated manually and not by chartsjs?

fanthos commented:
The chartjs one draws legend on canvas and hard to calculate chart height.

deel77 commented:
@fanthos I just want to ask, if it was some specific requirement to use "stepped" graphs (http://www.chartjs.org/samples/latest/charts/line/stepped.html)? Is there a way to change it?

fanthos commented:
@deel77 I am using the "stepped" to replace "interpolate" on Google charts and try to keep the lines looks similar as before.
According the code, the line will draw stepped if the data belongs a "sensor" domain.

balloob commented:
The problem is that we don't know if the value of a sensor can be interpolated.

In Home Assistant, a state will only change if the value is different. The exact same state will be treated as a new state if the entity property `force_update` is set to True. This allows us to interpolate the data from this entity. However, we don't record that this entity has force_update set to True, so we don't know
